phases/ephemeral/witx/wasi_ephemeral_sock.witx

```diff
@@ -11,34 +11,241 @@
+  (@interface func (export "close")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    (result $error $errno)
+  )
+
+  ;;; Enable/disable address reuse on a socket
+  ;;; Note: This is similar to `setsockopt` in POSIX for SO_REUSEADDR
+  (@interface func (export "set_reuse_addr")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    ;;; 1 to enable, 0 to disable
+    (param $reuse u8)
+    (result $error $errno)
+  )
+
+  ;;; Retrieve status of address reuse on a socket
+  ;;; Note: This is similar to `getsockopt` in POSIX for SO_REUSEADDR
+  (@interface func (export "get_reuse_addr")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    (result $error $errno)
+    (result $reuse u8)
+  )
+
+  ;;; Enable port reuse on a socket
+  ;;; Note: This is similar to `setsockopt` in POSIX for SO_REUSEPORT
+  (@interface func (export "set_reuse_port")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    ;;; 1 to enable, 0 to disable
+    (param $reuse u8)
+    (result $error $errno)
+  )
+
+  ;;; Retrieve status of port reuse on a socket
+  ;;; Note: This is similar to `getsockopt` in POSIX for SO_REUSEPORT
+  (@interface func (export "get_reuse_port")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    (result $error $errno)
+    (result $reuse u8)
+  )
+
+  ;;; Set size of receive buffer
+  ;;; Note: This is similar to `setsockopt` in POSIX for SO_RCVBUF
+  (@interface func (export "set_recv_buf_size")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    ;;; Buffer size
+    (param $size $size)
+    (result $error $errno)
+  )
+
+  ;;; Retrieve the size of the receive buffer
+  ;;; Note: This is similar to `getsockopt` in POSIX for SO_RCVBUF
+  (@interface func (export "get_recv_buf_size")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    (result $error $errno)
+    (result $size $size)
+  )
+
+  ;;; Set size of send buffer
+  ;;; Note: This is similar to `setsockopt` in POSIX for SO_SNDBUF
+  (@interface func (export "set_send_buf_size")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    ;;; Buffer size
+    (param $size $size)
+    (result $error $errno)
+  )
+
+  ;;; Retrieve the size of the send buffer
+  ;;; Note: This is similar to `getsockopt` in POSIX for SO_SNDBUF
+  (@interface func (export "get_send_buf_size")
+    ;;; Socket descriptor
+    (param $fd $fd)
+    (result $error $errno)
+    (result $size $size)
+  )
+
+  ;;; Bind a socket
+  ;;; Note: This is similar to `bind` in POSIX using PF_INET
+  (@interface func (export "bind")
+    ;;; File descriptor of the socket to be bind
+    (param $fd $fd)
+    ;;; Address to bind the socket to
+    (param $addr (@witx pointer $addr))
+    (result $error $errno)
+  )
+
+  ;;; Listen for connections on a socket
+  ;;; Note: This is similar to `listen`
+  (@interface func (export "listen")
+    ;;; File descriptor of the socket to be bind
+    (param $fd $fd)
+    ;;; Maximum size of the queue for pending connections
+    (param $backlog $size)
+    (result $error $errno)
+  )
```

Comment:
How useful are the server related functions (bind, listen, accept) for now?

Wasm is currently single-threaded and child connections can't be delegated to threads.